### PR TITLE
Adds priv_getcode

### PR DESCRIPTION
**Pull request by josh-richardson**
priv_getCode was previously implemented on another branch, but that became stale & difficult to rebase, so it was quicker to just branch off master and pull in my relevant commits to this branch by cherry-picking.

Signed-off-by: Joshua Richardson <joshua@richardson.tech>

## PR description
Adds priv_getCode, the implementation of eth_getCode but for private transactions.

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  private static final Logger LOG = getLogger();
+
+  private final BlockchainQueries blockchain;
+  private final PrivateStateRootResolver privateStateRootResolver;
+  private final PrivacyParameters privacyParameters;
+
+  public PrivGetCode(
+      final BlockchainQueries blockchain,
+      final PrivacyParameters privacyParameters,
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java]: It looks like we are passing down the `PrivacyParameter` object only to get an instance of the `privateStateArchive`. I think we should change this to receive the privateStateArchive directly.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @Override
+  public String getName() {
+    return RpcMethod.PRIV_GET_CODE.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    LOG.trace("Executing {}", RpcMethod.PRIV_GET_CODE.getMethodName());
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java]: We already have DEBUG log level for JSON-RPC calls. I don't think we need this one.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import org.hyperledger.besu.ethereum.core.PrivacyParameters;
+import org.hyperledger.besu.ethereum.privacy.PrivateStateRootResolver;
+
+import java.util.Optional;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class PrivGetCode implements JsonRpcMethod {
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java]: We have an abstraction for methods that have the block paramater to handle named or numbers, etc.
Take a look at `PrivCall` and how it extends `AbstractBlockParameterMethod`. I believe we should make `PrivGetCode` extend `AbstractBlockParameterMethod`.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  @Override
+  public String getName() {
+    return RpcMethod.PRIV_GET_CODE.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    LOG.trace("Executing {}", RpcMethod.PRIV_GET_CODE.getMethodName());
+
+    final Address address =
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java]: We have standardized that for any priv_ methods that have an equivalent eth_ method, we will make the first parameter (index 0) the privacyGroupId. And all other parameters will be the same.

So, the expected params for `priv_getCode` are:
```
params: [
   'privacy_group_id',
   '0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b',
   '0x2'  // 2
]
```

Take a look at `PrivCall` implementation

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        requestContext.getRequiredParameter(1, BlockParameter.class);
+
+    final String privacyGroupString = requestContext.getRequiredParameter(2, String.class);
+
+    final Bytes32 privacyGroupId = Bytes32.wrap(Bytes.fromBase64String(privacyGroupString));
+
+    final Hash latestStateRoot =
+        privateStateRootResolver.resolveLastStateRoot(
+            privacyGroupId,
+            blockParameter.isNumeric() && blockParameter.getNumber().isPresent()
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java]: As I mentioned before, extending `AbstractBlockParameterMethod` should help you handling the block paramter.

**Comment by lucassaldanha** [comment]
Would you please add an entry on `CHANGELOG.md` under the `1.4.1` section mentioning that we are adding `priv_getCode`?

**Comment by josh-richardson** [comment]
I've now addressed all comments from @lucassaldanha and @pinges, however by changing the parameter order, Web3j has had to undergo a refactor which has not yet been released, and therefore the `DeployPrivateSmartContractAcceptanceTest` will currently fail due to Web3j dependency sending the params in the incorrect order. I hope for this to be fixed on Monday & to be able to make another commit updating the Web3j dependency.

**Comment by josh-richardson** [comment]
Bumped the Web3j version, so this should now be ready to go.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCodeTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    when(mockArchive.get(lastStateRoot)).thenReturn(Optional.of(mockWorldState));
+    when(mockWorldState.get(resultantContractAddress)).thenReturn(mockAccount);
+    when(mockAccount.getCode()).thenReturn(fakeAccountCode);
+
+    final JsonRpcRequestContext request =
+        new JsonRpcRequestContext(
+            new JsonRpcRequest(
+                "2.0",
+                "priv_getCode",
+                new Object[] {
```
Comment (lucassaldanha) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCodeTest.java]: We need to fix the order of the parameters here.

**Pull request closed without merging**